File: docker/prepare-guide-media.R
#!/usr/bin/env Rscript

`%||%` <- function(x, y) if (is.null(x) || length(x) == 0L) y else x

lock_path <- file.path("desktop", "guide-media-lock.json")
if (!file.exists(lock_path)) stop("Missing CGV Guide media lock: ", lock_path, call. = FALSE)
if (!requireNamespace("jsonlite", quietly = TRUE)) stop("jsonlite is required to prepare CGV Guide media.", call. = FALSE)

lock <- jsonlite::fromJSON(lock_path)
expected <- sort(as.character(lock$expectedFiles %||% character(0)))
if (!identical(as.integer(lock$version), 1L) ||
    !grepl("^https://", as.character(lock$url), perl = TRUE) ||
    !grepl("^[a-f0-9]{64}$", as.character(lock$sha256), perl = TRUE) ||
    length(expected) != 35L) {
  stop("Invalid CGV Guide media lock.", call. = FALSE)
}

destination <- file.path("www", "screencasts")
is_complete <- dir.exists(destination) && all(file.exists(file.path(destination, expected)))
if (is_complete) {
  message("guide-media-ok files=", length(expected), " source=existing")
  quit(status = 0L)
}

archive <- tempfile("cgv-guide-media-", fileext = ".zip")
staging <- tempfile("cgv-guide-media-staging-")
on.exit(unlink(c(archive, staging), recursive = TRUE, force = TRUE), add = TRUE)
utils::download.file(as.character(lock$url), archive, mode = "wb", quiet = FALSE)

hash_line <- system2("sha256sum", archive, stdout = TRUE, stderr = TRUE)
actual_hash <- tolower(strsplit(trimws(hash_line[[1L]]), "\\s+", perl = TRUE)[[1L]][[1L]])
if (!identical(actual_hash, tolower(as.character(lock$sha256)))) {
  stop("CGV Guide media checksum mismatch.", call. = FALSE)
}

archive_entries <- utils::unzip(archive, list = TRUE)
entry_names <- sort(as.character(archive_entries$Name %||% character(0)))
if (!identical(entry_names, expected) || any(grepl("[/\\\\]", entry_names))) {
  stop("CGV Guide media archive contents do not match the lock.", call. = FALSE)
}

dir.create(staging, recursive = TRUE, showWarnings = FALSE)
utils::unzip(archive, exdir = staging)
sizes <- file.info(file.path(staging, expected))$size
if (any(!is.finite(sizes) | sizes <= 1024)) {
  stop("CGV Guide media archive contains missing or empty videos.", call. = FALSE)
}

unlink(destination, recursive = TRUE, force = TRUE)
dir.create(dirname(destination), recursive = TRUE, showWarnings = FALSE)
if (!file.rename(staging, destination)) stop("Could not install CGV Guide media.", call. = FALSE)
message("guide-media-ok files=", length(expected), " source=locked-release")

File: desktop/scripts/prepare-guide-media.js
const fs = require("fs");
const path = require("path");
const { downloadFile } = require("../src/download-file");
const { extractZipArchive, validateZipArchive } = require("../src/secure-zip");

const desktopRoot = path.resolve(__dirname, "..");
const repoRoot = path.resolve(desktopRoot, "..");
const lockPath = path.join(desktopRoot, "guide-media-lock.json");
const lock = JSON.parse(fs.readFileSync(lockPath, "utf8"));
const destination = path.join(repoRoot, "www", "screencasts");
const archivePath = path.join(desktopRoot, "resources", "cgv-guide-media-v1.zip");
const stagingPath = path.join(desktopRoot, "resources", "guide-media-staging");

function expectedFiles() {
  const files = Array.isArray(lock.expectedFiles) ? lock.expectedFiles.map(String).sort() : [];
  if (lock.version !== 1 || !/^https:\/\//.test(lock.url || "") || !/^[a-f0-9]{64}$/.test(lock.sha256 || "") || files.length < 30) {
    throw new Error("CGV Guide media lock is incomplete or invalid.");
  }
  if (files.some((name) => name.includes("/") || !name.endsWith(".mp4"))) {
    throw new Error("CGV Guide media lock contains an invalid filename.");
  }
  return files;
}

function mediaIsComplete(files) {
  return files.every((name) => {
    const target = path.join(destination, name);
    return fs.existsSync(target) && fs.statSync(target).isFile() && fs.statSync(target).size > 1024;
  });
}

async function main() {
  const files = expectedFiles();
  if (mediaIsComplete(files)) {
    process.stdout.write(`guide-media-ok files=${files.length} source=existing\n`);
    return;
  }

  fs.mkdirSync(path.dirname(archivePath), { recursive: true });
  await downloadFile(lock.url, archivePath, lock.sha256, (progress) => {
    if (progress.phase === "verifying") process.stdout.write("guide-media-verifying\n");
  });

  const archiveEntries = (await validateZipArchive(archivePath))
    .filter((name) => !name.endsWith("/"))
    .sort();
  if (JSON.stringify(archiveEntries) !== JSON.stringify(files)) {
    throw new Error("CGV Guide media archive contents do not match the locked file list.");
  }

  fs.rmSync(stagingPath, { recursive: true, force: true });
  fs.mkdirSync(stagingPath, { recursive: true });
  await extractZipArchive(archivePath, stagingPath);
  for (const name of files) {
    const target = path.join(stagingPath, name);
    if (!fs.existsSync(target) || fs.statSync(target).size <= 1024) {
      throw new Error(`CGV Guide media file is missing or empty: ${name}`);
    }
  }

  fs.rmSync(destination, { recursive: true, force: true });
  fs.renameSync(stagingPath, destination);
  process.stdout.write(`guide-media-ok files=${files.length} source=locked-release\n`);
}

main().catch((error) => {
  process.stderr.write(`${error.stack || error.message}\n`);
  process.exitCode = 1;
});
